src/meta/src/stream/stream_graph/assignment.rs

```diff
@@ -0,0 +1,2127 @@
+            // 2. Dividing by `w` biases selection towards containers with higher weight,
+            //    as smaller keys win in the rendezvous algorithm.
+            let key = -r.ln() / (weight.get() as f64);
+
+            match best {
+                None => best = Some((container, key)),
+                Some((_, best_key)) if key < best_key => best = Some((container, key)),
+                _ => {}
+            }
+        }
+
+        // quotas sum (possibly scaled) always >= items.len(), so best is always Some
+        let (winner, _) = best.expect("Invariant violation: no eligible container");
+        assignment
+            .entry(winner)
+            .and_modify(|v| v.push(item))
+            .or_insert_with(|| vec![item]);
+    }
+
+    assignment
+}
+
+/// Stable hash utility
+fn stable_hash<T: Hash>(t: &T) -> u64 {
+    let mut hasher = twox_hash::XxHash64::with_seed(0);
+    t.hash(&mut hasher);
+    hasher.finish()
+}
+
+/// A validated, non-negative, finite scale factor.
+#[derive(Debug, Copy, Clone)]
+pub struct ScaleFactor(f64);
+
+impl ScaleFactor {
+    /// Creates a new `ScaleFactor` if the value is valid.
+    ///
+    /// A valid scale factor must be finite and non-negative.
+    pub fn new(value: f64) -> Option<Self> {
+        if value.is_finite() && value >= 0.0 {
+            Some(ScaleFactor(value))
+        } else {
+            None
+        }
+    }
+
+    /// Gets the inner f64 value.
+    pub fn get(&self) -> f64 {
+        self.0
+    }
+}
+
+/// A no-op capacity scaling function: always returns `None`.
+pub fn unbounded_scale<C, I>(
+    _containers: &BTreeMap<C, NonZeroUsize>,
+    _items: &[I],
+) -> Option<ScaleFactor> {
+    None
+}
+
+/// A unit capacity scaling function: always returns `Some(1.0)`.
+pub fn weighted_scale<C, I>(
+    _containers: &BTreeMap<C, NonZeroUsize>,
+    _items: &[I],
+) -> Option<ScaleFactor> {
+    ScaleFactor::new(1.0)
+}
+
+/// Defines the capacity assignment strategy for containers.
+///
+/// - `Weighted`: Distribute items proportionally to container weights, applying any configured scale factor.
+/// - `Unbounded`: No capacity limit; containers can receive any number of items.
+#[derive(Clone, Copy, Debug)]
+#[non_exhaustive]
+pub enum CapacityMode {
+    /// Use each container’s weight to bound how many items it can receive.
+    /// When used for actor-to-worker assignment, this typically means actors are distributed
+    /// strictly proportionally to worker weights (i.e., using a scale factor of 1.0).
+    Weighted,
+
+    /// Ignore per-container quotas entirely—every container can take an unlimited number of items.
+    Unbounded,
+}
+
+/// Defines the vnode distribution strategy for hierarchical assignment.
+///
+/// - `RawWorkerWeights`: Distribute vnodes across workers using the original worker weight values.
+/// - `ActorCounts`: Distribute vnodes based on the number of actors assigned to each worker.
```

Comment:
Am I understanding it correctly: if there are a lot of workers/actors in the cluster, the results for different variant should be quite similar in theory, since the number of actors assigned should follow the weights?